File: 0331-verify-preorder-serialization-of-a-binary-tree/0331-verify-preorder-serialization-of-a-binary-tree.java
class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;
    public TreeNode(int val) {
        this.val = val;
    }
}
class Solution {
    Queue<String> queue;
    boolean ans;
    public boolean isValidSerialization(String preorder) {
        this.queue = new LinkedList<>(new ArrayList<>(Arrays.asList(preorder.split(","))));
        this.ans = true;
        
        build();
        return ans && queue.isEmpty();
    }

    TreeNode build() {
        if (queue.isEmpty()) {
            ans = false;
            return null;
        }

        String val = queue.poll();
        if (val.equals("#")) {
            return null;
        }

        TreeNode node = new TreeNode(Integer.valueOf(val));
        node.left = build();
        node.right = build();
        return node;
    }
}

File: 0331-verify-preorder-serialization-of-a-binary-tree/0331-verify-preorder-serialization-of-a-binary-tree.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
    
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        queue = collections.deque(preorder.split(","))
        self.ans = True

        def build():
            if not queue:
                self.ans = False
                return None
            
            ch = queue.popleft()
            if ch == "#":
                return None

            node = TreeNode(ch)
            node.left = build()
            node.right = build()
            return node
        
        build()
        return self.ans and not queue